### PR TITLE
Monolog handler

### PR DESCRIPTION
**Pull request by brettmc**
initial implementation of a logs bridge api for opentelemetry logging

### CODE REVIEW

File: src/Logs/Monolog/src/Handler.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+namespace OpenTelemetry\Contrib\Logs\Monolog;
+
+use Monolog\Formatter\NormalizerFormatter;
+use Monolog\Handler\AbstractHandler;
+use Monolog\Level;
+use Monolog\LogRecord;
+use OpenTelemetry\API\Common\Instrumentation\Globals;
+use OpenTelemetry\API\Logs as API;
+
+class Handler extends AbstractHandler
```
Comment (Nevay) [inline src/Logs/Monolog/src/Handler.php:14]: Could extend `AbstractProcessingHandler`.

File: src/Logs/Monolog/src/Handler.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        if (!$this->isHandling($record)) {
+            return false;
+        }
+        $normalized = $this->normalizer->format($record);
+        $logRecord = (new API\LogRecord())
+            ->setTimestamp($record->datetime->getTimestamp() * API\LogRecord::NANOS_PER_SECOND)
+            ->setSeverityNumber(API\Map\Psr3::severityNumber($record->level->toPsrLogLevel()))
+            ->setSeverityText($record->level->toPsrLogLevel())
+            ->setBody($normalized['message'])
+            ->setAttributes(array_merge($normalized['context'], $normalized['extra']))
```
Comment (Nevay) [inline src/Logs/Monolog/src/Handler.php:41]:
```suggestion
            ->setAttributes($normalized['context'] + $normalized['extra'])
```
IMO `context` should take precedence over `extra`.

File: src/Logs/Monolog/src/Handler.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+class Handler extends AbstractHandler
+{
+    private API\LoggerInterface $logger;
+    private NormalizerFormatter $normalizer;
+
+    /**
+     * @psalm-suppress ArgumentTypeCoercion
+     */
+    public function __construct(API\LoggerProviderInterface $loggerProvider = null, int|string|Level $level = Level::Debug, bool $bubble = true)
```
Comment (Nevay) [inline src/Logs/Monolog/src/Handler.php:22]: Would prefer if `$loggerProvider` is a required argument.

File: src/Logs/Monolog/src/Handler.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        parent::__construct($level, $bubble);
+        $this->normalizer = new NormalizerFormatter();
+        $loggerProvider ??= Globals::loggerProvider();
+        $this->logger = $loggerProvider->getLogger('monolog');
+    }
+
+    public function handle(LogRecord $record): bool
```
[the remark this answers is not in the document]
Comment (brettmc) [reply src/Logs/Monolog/src/Handler.php]: I've been able to make it work back to monolog v1.1 and php 7.4, which is almost everything. Prior to 1.1, composer didn't support autoloading.

File: src/Logs/Monolog/src/Handler.php (hunk 2 above)
Comment (brettmc) [reply src/Logs/Monolog/src/Handler.php]: done, and added a test for this

File: src/Logs/Monolog/src/Handler.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+     * @phan-suppress PhanTypeMismatchArgument
+     * @psalm-suppress InvalidOperand
+     */
+    protected function write($record): void
+    {
+        $logRecord = (new API\LogRecord())
+            ->setTimestamp((int) ($record['datetime']->format('Uu') * 1000))
+            ->setSeverityNumber(API\Map\Psr3::severityNumber($record['level_name']))
+            ->setSeverityText($record['level_name'])
+            ->setBody($record['message'])
```
Comment (Nevay) [inline src/Logs/Monolog/src/Handler.php:33]:
```suggestion
            ->setBody($record['formatted'])
```
and change `::getDefaultFormatter()` to a formatter that returns `$record['message']` (could use `new LineFormatter('%message%')` if we do not care about the unnecessary normalization) to allow users to change the body format.

**Review by Nevay — COMMENTED** [review]
We should also include `$record['channel']` as attribute.
Would it make sense to record `context`, `extra` and `channel` as separate attributes to prevent overriding data?

**Comment by brettmc** [comment]
> Would it make sense to record context, extra and channel as separate attributes to prevent overriding data?

I've put them all into their own attributes (and set the default formatter to `NormalizerFormatter` so that exceptions can be logged). I think let's get a 0.0.1 release out, and see whether the log record structure is useful when people actually use it.

File: src/Logs/Monolog/src/Handler.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+     */
+    public function __construct(API\LoggerProviderInterface $loggerProvider, $level, bool $bubble = true)
+    {
+        parent::__construct($level, $bubble);
+        $this->logger = $loggerProvider->getLogger('monolog');
+    }
+
+    protected function getDefaultFormatter(): FormatterInterface
+    {
+        return new NormalizerFormatter('Uu');
```
Comment (Nevay) [inline src/Logs/Monolog/src/Handler.php:27]:
```suggestion
        return new NormalizerFormatter();
```
I don't think we should change the dateFormat as it affects datetimes in `context` and `extra` too; I think we should populate timestamp, severity and channel from the original record and use the formatted record for body, context and extra.

**Review by bobstrecansky — APPROVED** [review]

**Pull request merged**